libsel4utils/src/process.c

```diff
@@ -271,9 +271,22 @@ int sel4utils_spawn_process_v(sel4utils_process_t *process, vka_t *vka, vspace_t
     int error;
     int envc = 0;
     char *envp[] = {};
+    Elf_Phdr *phdr = process->elf_phdrs;
 
     uintptr_t initial_stack_pointer = (uintptr_t) process->thread.stack_top - sizeof(seL4_Word);
 
+    /* If PT_PHDR exists in the program headers, assign PT_NULL to it.
+     * This is because muslc libc searches for PT_PHDR and if found,
+     * it assumes it's part of the ELF image and relocates the entire
+     * subsequent program header segments according to PT_PHDR's base. This is
+     * wrong and will trigger mapping errors.
+     */
+    for (int i = 0; i < process->num_elf_phdrs; i++, phdr++) {
+        if (phdr->p_type == PT_PHDR) {
+            phdr->p_type = PT_NULL;
+        }
+    }
+
```

Comment:
The ELF documentation says:

> it may occur only if the program header table is part of the memory image of the program.

So I agree that this should be done.

I do think this is the wrong place though. You either have to skip the `PT_PHDR` copy in `sel4utils_elf_read_phdrs()` once, or do it on the stack copy below.

Or at least do this loop after the `sel4utils_elf_read_phdrs()` call in `sel4utils_configure_process_custom()`.

Reply:
Can't do it in the stack copy as it doesn't know what it's copying, also cannot after the stack copy as it's not mapped anymore. I added it in `sel4utils_configure_process_custom()` just after `sel4utils_elf_read_phdrs`. `sel4utils_elf_read_phdrs` might be called by other code/libraries that may need to do something different with `PT_PHDR`?